airbyte-webapp/src/core/form/schemaToYup.test.ts

```diff
@@ -0,0 +1,269 @@
+      is_field_no_default: {
+        type: "boolean",
+      },
+      dbname: { type: "string", description: "Name of the database." },
+      password: {
+        type: "string",
+        description: "Password associated with the username.",
+      },
+      reports: {
+        type: "array",
+        items: {
+          type: "string",
+        },
+      },
+    },
+    additionalProperties: false,
+  };
+  const yupSchema = buildYupFormForJsonSchema(schema, jsonSchemaToFormBlock(schema));
+
+  const expectedSchema = yup.object().shape({
+    host: yup.string().trim().required("form.empty.error").transform(String),
+    port: yup
+      .number()
+      .min(0)
+      .max(65536)
+      .required("form.empty.error")
+      .transform((val) => (isNaN(val) ? undefined : val)),
+    user: yup.string().trim().required("form.empty.error").transform(String),
+    is_sandbox: yup.boolean().default(false),
+    is_field_no_default: yup.boolean().required("form.empty.error"),
+    dbname: yup.string().trim().required("form.empty.error").transform(String),
+    password: yup.string().trim().transform(String),
+    reports: yup.array().of(yup.string().trim().transform(String)),
+  });
+
+  expect(JSON.stringify(yupSchema)).toEqual(JSON.stringify(expectedSchema));
+});
+
+const simpleConditionalSchema: AirbyteJSONSchema = {
+  type: "object",
+  required: ["start_date", "credentials"],
+  properties: {
+    start_date: {
+      type: "string",
+    },
+    max_objects: {
+      type: "number",
+    },
+    credentials: {
+      type: "object",
+      oneOf: [
+        {
+          title: "api key",
+          required: ["type", "api_key"],
+          properties: {
+            api_key: {
+              type: "string",
+              pattern: "\\w{5}",
+            },
+            type: {
+              type: "string",
+              const: "api",
+              default: "api",
+            },
+          },
+        },
+        {
+          title: "oauth",
+          required: ["type", "redirect_uri"],
+          properties: {
+            redirect_uri: {
+              type: "string",
+              examples: ["https://api.hubspot.com/"],
+            },
+            type: {
+              type: "string",
+              const: "oauth",
+              default: "oauth",
+            },
+          },
+        },
+      ],
+    },
+  },
+};
+
+it("should build  mixed schema for conditional case", () => {
+  const yupSchema = buildYupFormForJsonSchema(simpleConditionalSchema, jsonSchemaToFormBlock(simpleConditionalSchema));
+
+  const expectedSchema = yup.object().shape({
+    start_date: yup.string().trim().required("form.empty.error").transform(String),
+    max_objects: yup.number().transform((x) => x),
+    credentials: yup.object().shape({
+      type: yup.mixed(),
+      api_key: yup
+        .mixed()
+        .when("type", { is: "", then: (x) => x, otherwise: (x) => x })
+        .when("type", { is: "", then: (x) => x, otherwise: (x) => x }),
```

Comment:
This doesn't make sense to me. Wouldn't we expect the `api_key` schema to be something like
```
api_key: yup
  .mixed()
  .when("type", { is: "api", then: (x) => x.string().matches(\\w{5}), otherwise: (x) => x })
  .when("type", { is: (val) => !["api"].includes(val), then: (x) => x.strip() })
```

same comment goes for the `redirect_uri` schema below, and the schemas at the end of this file

Reply:
This test is only testing the structure of the yup schema which does not capture the actual conditions of the whens:
```json
 "api_key": {
              "type": "mixed",
              "_whitelist": {
                "list": {},
                "refs": {}
              },
              "_blacklist": {
                "list": {},
                "refs": {}
              },
              "exclusiveTests": {},
              "deps": [
                "type",
                "type"
              ],
              "conditions": [
                {
                  "refs": [
                    {
                      "key": "type",
                      "isContext": false,
                      "isValue": false,
                      "isSibling": true,
                      "path": "type"
                    }
                  ]
                },
                {
                  "refs": [
                    {
                      "key": "type",
                      "isContext": false,
                      "isValue": false,
                      "isSibling": true,
                      "path": "type"
                    }
                  ]
                }
              ],
```

The tests below this one (the stuff in `should schema build conditional case that validates correctly based on selection key`) test whether the `when`s actually kick in. 

Added a comment to explain better.